examples/aggregator/common/src/lib.rs

```diff
@@ -43,34 +56,31 @@ impl<T: Monoid> ThresholdAggregator<T> {
             current_value: Monoid::identity(),
             sample_count: 0,
             sample_threshold: threshold,
+            exhausted: false,
         }
     }
 
     /// Combines a new sample with the current aggregated value.
-    pub fn submit(&mut self, sample: &T) {
-        self.current_value = self.current_value.combine(sample);
-        self.sample_count += 1;
-    }
-
-    /// Returns the current aggregated value if the number of collected samples is at least
-    /// `sample_threshold`.
-    pub fn get(&self) -> Option<&T> {
-        if self.sample_count >= self.sample_threshold {
-            Some(&self.current_value)
-        } else {
-            None
-        }
-    }
-
-    /// If the number of current samples is at least `sample_threshold`, then returns the current
-    /// aggregated value and resets it; otherwise, returns `None` and leaves the internal state
-    /// unchanged.
-    pub fn take(&mut self) -> Option<T> {
-        if self.sample_count >= self.sample_threshold {
-            self.sample_count = 0;
-            Some(replace(&mut self.current_value, Monoid::identity()))
+    ///
+    /// If the number of current samples (including the current one) is exactly `sample_threshold`,
+    /// then returns the current aggregated value and sets the status of the aggregator to
+    /// exhausted.
+    ///
+    /// If the aggregator was already exhausted, the new sample is discarded.
+    pub fn submit(&mut self, sample: &T) -> AggregatorResult<&T> {
+        if self.exhausted {
+            AggregatorResult::Exhausted
         } else {
-            None
+            self.current_value = self.current_value.combine(sample);
+            self.sample_count += 1;
+            if self.sample_count >= self.sample_threshold {
+                // We set the exhausted flag, so that future calls to this method will not return
+                // additional values, since that would reveal the actual value of new samples.
+                self.exhausted = true;
+                AggregatorResult::AggregatedValue(&self.current_value)
+            } else {
+                AggregatorResult::BelowThreshold
+            }
```

Comment:
Ah that makes sense. :)) I was wondering why this PR seems to change behaviour but this explains the issue it fixes.

Nit, but might be nice to make this clearer in the description? Based off the description I thought this was a refactor only.